### DISCUSSION TITLE
CriteriaBuilder with embedded properties

### DISCUSSION THREAD

**Discussion by alejandropg** [micronaut-projects/micronaut-data#1345] [Q&A]
Hi,

I'm trying to use `CriteriaBuilder` to build criteria to find some entities by a property of the embedded object.

An example:

```kotlin
@MappedEntity
data class Book(
    @field:Id @GeneratedValue val id: Int,
    val name: String,

    @Relation(EMBEDDED)
    val parent: BookParent
)

@Embeddable
data class BookParent(
    val type: String,
    val id: Int
) {
```

I tried (well I tried several things but I suppose that the more "correct" is this one, and I know a can do it whit a simple compile-time query, but this is just a simpler example to show my problem):

```kotlin
object BookPredicates {

    fun byParentId(parentId: Int): PredicateSpecification<Book> =
        PredicateSpecification<Book> { root, cb ->
            val parentPath = root.get<BookParent>("parent")
            val parentIdPath = parentPath.get<Int>("id")

            cb.equal(parentIdPath, parentId)
        }

}
```

With this approach I get the exception `IllegalStateException("Not supported operation!")` which is throw by the class `io.micronaut.data.model.jpa.criteria.impl.AbstractPersistentPropertyPath<T>` in the method:

```java
    @Override
    public <Y> Path<Y> get(String attributeName) {
        throw notSupportedOperation();
    }
```

I'm doing something wrong or just this feature of navigating to related entities is not supported?

Thanks a lot! 🙏

**Comment by dstepanov** [comment]
I would expect it to work, Criteria API has a lot of different use-cases so this might be a missed one.

**Reply by alejandropg** [reply]
ok, no problem.

May I help in any way?

**Reply by alejandropg** [reply]
By the way, to try to be more specific, the exception is throw in the line `val parentIdPath = parentPath.get<Int>("id")`.

I mean, the first `get` works fine and I get the path of "parent" property, but the exception of "not supported operation" is when I call the `parentPath.get<Int>("id")`.

**Reply by dstepanov** [reply]
It looks like there needs to be a variation of `RuntimePersistentPropertyPathImpl` for `RuntimeEmbedded` that implements the class `RuntimePersistentEntityPath`

**Reply by alejandropg** [reply]
But, I should this change in my code? or it's an internal issue?

**Reply by dstepanov** [reply]
It's internal, you can try to implement it

**Comment by alejandropg** [comment]
Uffff I would like and thanks for your confidence, but I think this task is very far from my knowledge of Micronaut Data for now 😅🙈

I'm working go deeper in the code, but It will take me time 😔
